### DISCUSSION TITLE
[Feature] "{album}" in storage template

### DISCUSSION THREAD

**Discussion by Hely0n** [immich-app/immich#2933] [Feature Request]
### The feature

I fell in love with immich and I'm sure it will be our lifetime Memory management. The only thing I'm waiting for, as many others, is the Folder watch feature, or in other words, the possibility to use an existing folder structure. 

But then I tought, why not just bulk import anything with the CLI tool. The files aren't touched and I can specify the folder structure and it even supports creating albums per foldername! So lets go!

But the problem is, that there is no Template for "Album" which will inevitably end in a mess on the file system. We, as users, have to think about the case, where immich will be abandoned or there is another reason that we have to stop using it sadly. So filewise, the important information of "to which event does these photos and videos belong", must be intact on the file system.

My Expectation is, to just add an {album} template to the possible storage template.

- If the element doesn't belong to any album, the folder will be named "unsorted" or something.
- If the element belongs to multiple albums, it will choose the first album it was added (like the main album). Or split it in {firstAlbum} and {lastAlbum}

I will look at the code and see, if I'm mighty enough to implement it myself and make a pull request. But I can't promise anything :D 

### Platform

- [X] Server
- [X] Web
- [ ] Mobile

**Comment by alextran1502** [comment]
FYI, limited support for read-only existing gallery is out since the last release. You can read more here https://immich.app/docs/features/read-only-gallery

**Reply by Hely0n** [reply]
Yes, I saw it and that would make a great workaround in the meantime. But of course just a temporary due to the limitation. As I said, I'm ready to pipe my whole gallery through immich once and it should't be read-only necessarily (I have backups for the worst case), but it must preserve the original folder/album names after the import

**Reply by Hely0n** [reply]
Could you please point me to the code of the storage templates? I want to look at it

**Reply by alextran1502** [reply]
You can find it in this folder `server/src/domain/storage-template`

**Reply by Hely0n** [reply]
Okay thank you. Well that looks very complex at the first look, but would't it do it, if the option will be added to the system-configs and and to modify render() (i think) to interpret the keyword accordingly? (so aquiring the first album name of the element) ?

**Comment by Hely0n** [comment]
So I managed to implement it basically with this one-liner in tthe render()-Method

`substitutions['album'] = Array.isArray(asset.albums) && asset.albums.length > 0 ? asset.albums[0].albumName : "." //just throw into the root if it doesn't belong to an album`

It kind of works. If I use "unsorted" instead of the ".", it creates a subfolder called "unsorted", as it should, when uploading.

But the problem is, that adding the photo to the album afterwards and running the storage migration, doesn't change the location. I think the recognition of a changed album must me implemented in the storage migration job.

Could you please point me to the code position, where this is handled or explain, how the process works? @alextran1502

**Reply by Hely0n** [reply]
Okay it looks like the migration job does indeed recognize the changed path, but it still behaves like the assets doesn't belong to an album, even if they do

**Reply by Hely0n** [reply]
Okay it seems like assets.albums is just undefined and I don't understand why

**Reply by Hely0n** [reply]
I tried to add "album: True" to getById() here: https://github.com/immich-app/immich/blob/main/server/src/infra/repositories/asset.repository.ts#L66

But it's still undefined

**Comment by Hely0n** [comment]
#2973
